### PR TITLE
CSS Backgrounds and Borders 3 parsing tests

### PR DESCRIPTION
**Pull request by ewilligers**
Test that each property accepts valid values and
rejects invalid values.

Spec: https://drafts.csswg.org/css-backgrounds/

### CODE REVIEW

**Comment by w3c-bots** [comment]
# Build PENDING

Started: None
Finished: None





View more information about this build on:

- [WPT PR Status](http://pulls.web-platform-tests.org/build/26427)
- [TravisCI](https://travis-ci.org/w3c/web-platform-tests/builds/358044230)

**Comment by ewilligers** [comment]
@emilio  please review.

Browser bugs have been reported:
https://developer.microsoft.com/en-us/microsoft-edge/platform/issues/16555303/
https://developer.microsoft.com/en-us/microsoft-edge/platform/issues/16575199/
https://developer.microsoft.com/en-us/microsoft-edge/platform/issues/16853261/

https://bugs.webkit.org/show_bug.cgi?id=183990
https://bugs.webkit.org/show_bug.cgi?id=183993
https://bugs.webkit.org/show_bug.cgi?id=183994

https://bugs.chromium.org/p/chromium/issues/detail?id=825549
https://bugs.chromium.org/p/chromium/issues/detail?id=825568
https://bugs.chromium.org/p/chromium/issues/detail?id=817471

https://bugzilla.mozilla.org/show_bug.cgi?id=1448646
https://bugzilla.mozilla.org/show_bug.cgi?id=1448656

File: css/css-backgrounds/parsing/border-image-outset-valid.html [hunk 1]
```diff
@@ -12,8 +12,10 @@
 </head>
 <body>
 <script>
-test_valid_value("border-image-outset", "1px");
-test_valid_value("border-image-outset", "2");
+// "1px" in Blink, Edge, WebKit; "1px 1px 1px 1px" in Firefox.
+test_valid_value("border-image-outset", "1px", ["1px", "1px 1px 1px 1px"]);
```
Comment (emilio) [inline css/css-backgrounds/parsing/border-image-outset-valid.html]: FWIW, I don't know with which Firefox version have you tested, but Firefox serializes to 1px for me.

File: css/css-backgrounds/parsing/border-image-outset-valid.html [hunk 2]
```diff
@@ -12,8 +12,10 @@
...
+// "2" in Blink, Edge, WebKit; "2 2 2 2" in Firefox.
+test_valid_value("border-image-outset", "2", ["2", "2 2 2 2"]);
```
Comment (emilio) [inline css/css-backgrounds/parsing/border-image-outset-valid.html]: Similarly, this serializes to `2` for me.

File: css/css-backgrounds/parsing/border-image-repeat-valid.html [hunk 3]
```diff
@@ -12,10 +12,11 @@
 </head>
 <body>
 <script>
-test_valid_value("border-image-repeat", "stretch");
-test_valid_value("border-image-repeat", "repeat round");
-
+// "stretch" in Blink, Edge, WebKit; "stretch stretch" in Firefox.
+test_valid_value("border-image-repeat", "stretch", ["stretch", "stretch stretch"]);
```
Comment (emilio) [inline css/css-backgrounds/parsing/border-image-repeat-valid.html]: This just serializes to `stretch` on FF Nightly.

File: css/css-backgrounds/parsing/border-image-repeat-valid.html [hunk 4]
```diff
@@ -12,10 +12,11 @@
...
+// "space" in Blink, Edge, WebKit; "space space" in Firefox.
 test_valid_value("border-image-repeat", "space space", ["space", "space space"]);
```
Comment (emilio) [inline css/css-backgrounds/parsing/border-image-repeat-valid.html]: This does serialize as `space space`, I'll fix, thanks!

File: css/css-backgrounds/parsing/border-image-valid.html [hunk 5]
```diff
@@ -14,35 +14,46 @@
...
-test_valid_value("border-image", "1 / / 1px");
-test_valid_value("border-image", "1 2% 3 4% / / 2");
-test_valid_value("border-image", "1 2% 3 4% fill / / 1px 2 3px 4");
-
-test_valid_value("border-image", "1 / 1px / 1px");
-test_valid_value("border-image", "1 2% 3 4% / 2% / 2");
-test_valid_value("border-image", "1 2% 3 4% fill / 3 / 1px 2 3px 4");
-test_valid_value("border-image", "1 / auto / 1px");
-test_valid_value("border-image", "1 2% 3 4% / 1px 2% 3 auto / 2");
+// "none" in Edge, "none 100% / 1 / 0 stretch" in Firefox.
```
Comment (emilio) [inline css/css-backgrounds/parsing/border-image-valid.html]: What about WK / Blink? I think Edge would be ideally right here.

File: css/css-backgrounds/parsing/background-color-invalid.html [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+<link rel="help" href="https://drafts.csswg.org/css-backgrounds/#background-color">
+<meta name="assert" content="background-color supports only the grammar '<color>'.">
+<script src="/resources/testharness.js"></script>
+<script src="/resources/testharnessreport.js"></script>
+<script src="resources/parsing-testcommon.js"></script>
+</head>
+<body>
+<script>
+test_invalid_value("background-color", "none");
+
```
Comment (emilio) [inline css/css-backgrounds/parsing/background-color-invalid.html]: nit: stray newline?

I'd also remove the `<html>` / `<head>` / `<body>` tags, to keep the test smaller, but...

File: css/css-backgrounds/parsing/background-valid.html [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+<link rel="help" href="https://drafts.csswg.org/css-backgrounds/#background">
+<meta name="assert" content="background supports the full grammar '<bg-layer># , <final-bg-layer>'.">
+<script src="/resources/testharness.js"></script>
+<script src="/resources/testharnessreport.js"></script>
+<script src="resources/parsing-testcommon.js"></script>
+</head>
+<body>
+<script>
+// For <final-bg-layer>, implementations serialize <‘background-color’> after <bg-image>, contrary to spec.
+// Safari removes quotes.
```
Comment (emilio) [inline css/css-backgrounds/parsing/background-valid.html]: Hmm, does safari remove quotes unconditionally? How does it handle escaping? That's maybe worth a bug.

File: css/css-backgrounds/parsing/border-image-repeat-valid.html (hunk 4 above)
Comment (emilio) [reply css/css-backgrounds/parsing/border-image-repeat-valid.html]: Opened https://github.com/servo/servo/pull/20610 for this.

File: css/css-backgrounds/parsing/border-image-slice-valid.html [hunk 8]
```diff
@@ -12,7 +12,8 @@
 </head>
 <body>
 <script>
-test_valid_value("border-image-slice", "1");
+// "1" in Blink, Edge, WebKit; "1 1 1 1" in Firefox.
+test_valid_value("border-image-slice", "1", ["1", "1 1 1 1"]);
```
Comment (emilio) [inline css/css-backgrounds/parsing/border-image-slice-valid.html]: FF nightly serializes `1` for me.

**Review by emilio — APPROVED** [review]
LGTM, but feel free to remove the longer serialization tests that FF serializes correctly, if you want.

**Comment by ewilligers** [comment]
I was testing Firefox using Firefox ESR 52.6.0 and Firefox 59. Now using Nightly only.

Additional serialization issues raised:
https://github.com/w3c/csswg-drafts/issues/2547
https://developer.microsoft.com/en-us/microsoft-edge/platform/issues/16905526/

File: css/css-backgrounds/parsing/border-image-valid.html (hunk 5 above)
Comment (ewilligers) [reply css/css-backgrounds/parsing/border-image-valid.html]: Blink and WebKit fail all these tests by returning an empty string as the value of border-image.

File: css/css-backgrounds/parsing/border-image-outset-valid.html (hunk 1 above)
Comment (ewilligers) [reply css/css-backgrounds/parsing/border-image-outset-valid.html]: I observed "1px 1px 1px 1px" using Firefox ESR 52.6.0

File: css/css-backgrounds/parsing/background-valid.html (hunk 7 above)
Comment (ewilligers) [reply css/css-backgrounds/parsing/background-valid.html]: https://github.com/w3c/csswg-drafts/issues/2547
https://bugs.webkit.org/show_bug.cgi?id=28869

**Pull request merged**